src/hooks/useFetch.js

```diff
@@ -4,103 +4,86 @@
-      (async () => {
-        const { signal } = abortController;
-        const headers = { authorization: getAuthorizationHeader() };
-        try {
-          const response = await fetch(url, { headers, signal, ...options });
-          const { ok, status } = response;
-          if (ok) {
-            const responseData = await response.json();
-            const { error: responseError } = responseData;
-            if (responseError) throw new BugsnagError(responseError, { url, headers });
-            if (!responseData.length) throw new Error(ERROR_CODES.EMPTY_RESPONSE);
-            if (isMounted.current) setData(responseData);
-          } else {
-            switch (status) {
-              case 400:
-              default:
-                throw new Error(ERROR_CODES.CONNECTION_FAILURE);
-              case 401:
-                throw new BugsnagError(ERROR_CODES.INVALID_PASSWORD, { url, headers });
-            }
-          }
-        } catch (responseError) {
-          if (isMounted.current) {
-            if (!isValidError(responseError)) {
-              responseError.message = ERROR_CODES.CONNECTION_FAILURE;
-            }
-            setError(responseError);
-          }
-        }
-      })();
-      const timer = setTimeout(() => {
-        if (isMounted.current) setData();
-        abortController.abort();
-      }, timeout);
-      return () => {
-        isMounted.current = false;
-        clearTimeout(timer);
-        abortController.abort();
-      };
-    }
+    _isMounted.current = true;
     return () => {
-      isMounted.current = false;
+      _isMounted.current = false;
+      if (_controller.current) _controller.current.abort();
+    };
+  }, []);
+
+  const _fetch = async (path, init, opts) => {
+    const { responseHandler = res => res, errorHandler = err => err, timeout = 10000 } = opts;
+
+    const onResponse = async res => setResponse(await responseHandler(res));
+    const onError = async err => setError(await errorHandler(err));
+
+    const beforeFetch = () => {
+      _isBlocked.current = true;
+      // eslint-disable-next-line no-undef
+      _controller.current = new AbortController();
+      _timer.current = setTimeout(() => {
+        if (_controller.current) _controller.current.abort();
+      }, timeout);
+      setIsLoading(true);
     };
-  }, [url]);
 
-  const isFetching = !(data || error);
-  return [data, error, isFetching];
+    const tryFetch = async () => fetch(url + path, { ...init, signal: _controller.current.signal });
+
+    const afterFetch = () => {
+      _isBlocked.current = false;
+      _controller.current = null;
+      clearTimeout(_timer.current);
+      setIsLoading(false);
+    };
+
+    const doFetch = async () => {
+      try {
+        beforeFetch();
+        const res = await tryFetch();
+        await onResponse(res);
+      } catch (err) {
+        await onError(err);
+      } finally {
+        afterFetch();
+      }
+    };
+
+    doIf(isUnblocked, doFetch);
```

Comment:
This is way clearer now, nice! My personal taste (and makes no difference whatsoever), just for the different perspective would be to not use the functions `beforeFetch`, `tryFetch` and `AfterFetch`, `doFetch` and to just have the code be written sequentially. Maybe it's a bit ugly but it's just so easy to read. At the moment I think I read it more similarly to not using any `async/await`, and just using `then` calls

Reply:
No worries, I'm not too bothered on `await` v. `then`, generally whichever is cleanest! I think this should be resolved now anyways with the simplication of the API to make the handlers both synchronous.